### PR TITLE
Upgrade CI clang-format to 6.0

### PR DESCRIPTION
**Pull request by raulchen**
## What do these changes do?

The travis CI job is currently using clang-format 3.8, which is too old and behaves differently with modern versions on developers' environment. 
This RP upgrades clang-format to 6.0, and format c++ code accordingly.
It also upgrades CI job's linux platform from Ubuntu 12.04 to 14.04. Because clang-format 6 cannot be installed on the old version.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 language: generic
+# Use Ubuntu 16.04
+dist: xenial
```
Comment (robertnishihara) [inline .travis.yml]: Should this be specifically in the `-os: linux` sections?

**Review by robertnishihara — APPROVED** [review]
LGTM except for my one comment.

File: ci/travis/check-git-clang-format-output.sh [hunk 2]
```diff
@@ -8,7 +8,7 @@ else
   base_commit="$TRAVIS_BRANCH"
   echo "Running clang-format against branch $base_commit, with hash $(git rev-parse $base_commit)"
 fi
-output="$(ci/travis/git-clang-format --binary clang-format-3.8 --commit $base_commit --diff --exclude '(.*thirdparty/|.*redismodule.h|.*.js|.*.java)')"
+output="$(ci/travis/git-clang-format --binary clang-format --commit $base_commit --diff --exclude '(.*thirdparty/|.*redismodule.h|.*.js|.*.java)')"
```
Comment (robertnishihara) [inline ci/travis/check-git-clang-format-output.sh]: Could be done in a separate PR, but we should throw an error if the version of `clang-format` is not the one we test in the CI.

Same with `flake8` and `yapf`, we should raise exceptions if the user has a different version (in the `ci/travis/format.sh` script).

File: .travis.yml (hunk 1 above)
Comment (raulchen) [reply .travis.yml]: Putting this here would apply to all linux jobs. Otherwise we'll have to add this for each of them.

Comment (robertnishihara) [reply .travis.yml]: Ok, that's fine.

**Comment by simon-mo** [comment]
@raulchen It seems like this PR failed the valgrind test. I'll try to reproduce it and pin point the issue

**Comment by raulchen** [comment]
> @raulchen It seems like this PR failed the valgrind test. I'll try to reproduce it and pin point the issue

Thanks. Did you find what was wrong? This PR only formatted code, didn't change any code logic.

**Comment by pcmoritz** [comment]
The upgrade to Ubuntu 16.05 makes valgrind fail, probably the new valgrind
version that ships with it is more strict or changed behavior.

**Pull request merged**